### PR TITLE
Clean-up symex assignments

### PR DESCRIPTION
**Pull request by romainbrenguier**
This introduce an SSA_assignment_stept class to clarify what describes an assignment step (this could be a start to progressively split the SSA_stept class into several more specialized ones) and does some simple clean-up. 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/symex_assign.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/exception_utils.h>
 #include <util/pointer_offset_size.h>
 #include <util/simplify_expr.h>
+#include <util/expr_util.h>
```
Comment (tautschnig) [inline src/goto-symex/symex_assign.cpp]: Please sort lexicographically (I'm pretty sure clang-format will block on this)

**Review by tautschnig — APPROVED** [review]

File: src/goto-symex/symex_assign.cpp [hunk 2]
```diff
@@ -485,12 +486,12 @@ void goto_symext::symex_assign_symbol(
               << messaget::eom;
     });
 
-  // Temporarily add the state guard
-  guard.emplace_back(state.guard.as_expr());
+  const exprt assignment_guard =
+    make_and(state.guard.as_expr(), conjunction(guard));
```
Comment (owen-mc-diffblue) [inline src/goto-symex/symex_assign.cpp:490]: Presumably this was done this way in the name of performance. Have you checked that your change doesn't noticeably slow symex down?

**Review by owen-mc-diffblue — APPROVED** [review]

**Comment by tautschnig** [comment]
@romainbrenguier This now needs a rebase, which I guess would also sort out the CI failure (it's a spurious clang-format report).

File: src/goto-symex/symex_assign.cpp (hunk 2 above)
Comment (romainbrenguier) [reply src/goto-symex/symex_assign.cpp:490]: Before this was adding an exprt to a vector then using this vector to make the `and_exprt` of the conjunction, now we first make the conjunction and then add the exprt to the vector of the `and_exprt` operands. So the operation are the same, just in a different order, this shouldn't impact performances.

Comment (romainbrenguier) [reply src/goto-symex/symex_assign.cpp:490]: @owen-jones-diffblue I just noticed that my change could lead to `assignment_guard` being `TRUE && TRUE` so I've added a commit to have `make_and` handle the constant cases, can you check that?

Comment (owen-mc-diffblue) [reply src/goto-symex/symex_assign.cpp:490]: It seems good, once I saw from `exprt::is_true()` that boolean constant `exprt`s must be either `true_exprt` or `false_exprt`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 63e01d9bf47bc49a3f4231927db3c6f5a0b51662).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114089912

**Pull request merged**